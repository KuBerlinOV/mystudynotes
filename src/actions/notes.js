import { v4 as uuidv4 } from 'uuid';

//ADD_NOTE
export const addNote = ({ topic = '', // in this case you should imagine destructuring as you are passing some object with the same data like ' const movie = {title, year, id} '
    description = '',           //when we are passing it to the action function as an argument and destructuring it. 
    note = '',                 //when it would be like ' const {title, year, id} = movie ' 
    reference = '',            //Furthermore, we provide here the default values for the object keys so
    tag = '#',
    status = 'in progress',     //the user could leave these options empty if he wanted to.
    createdAt = 0
} = {}) => {
    return {
        type: 'ADD_NOTE',
        note: {
            id: uuidv4(),
            topic,
            description,
            note,
            reference,
            tag,
            status,
            createdAt
        }
    }
}

//EDIT_NOTE

export const updateNote = (id, updates) => ({
    type: 'UPDATE_NOTE',
    id,
    updates
})

export const updateStatus = (id) => ({
    type: 'UPDATE_STATUS',
    id
})

//REMOVE_NOTE

export const removeNote = (id) => ({
    type: 'REMOVE_NOTE',
    id
})