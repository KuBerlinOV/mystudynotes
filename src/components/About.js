import React from 'react';

const About = () => {
    return (
        <div>organizing your studies</div>
    )
}

export default About;