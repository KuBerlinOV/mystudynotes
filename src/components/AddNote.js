import React from 'react';
import Modal from 'react-modal';

const AddNote = ({ topic, description, note, reference, tag, error, handleChange, handleSubmit, trigger }) => {


    return (trigger) ?
        (<div className='popup-addnote'>
            <h3 color="red">Note</h3>
            <form onSubmit={handleSubmit} action="">
                <h3>Topic:</h3>
                <input type="text" placeholder="topic" name="topic" autoFocus value={topic} onChange={handleChange} />
                <h3>Description</h3>
                <input type="text" placeholder="description" name="description" id="" value={description} onChange={handleChange} />
                <h3>Note</h3>
                <textarea placeholder="write your note here" name="note" id="" cols="30" rows="10" value={note} onChange={handleChange} ></textarea>
                <h3>Reference</h3>
                <input type="text" placeholder="reference" name="reference" id="" value={reference} onChange={handleChange} />
                <h3>Tag</h3>
                <input type="text" placeholder="tag" name="tag" value={tag} onChange={handleChange} />

                {error && <p>Please provide topic and description</p>}
                <button>Save</button>

            </form>
        </div>
        ) : ''
}

export default AddNote;