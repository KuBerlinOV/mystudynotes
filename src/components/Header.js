import React from 'react';
import { NavLink } from 'react-router-dom';


const Header = () => (
    <header>
        <NavLink activeClassName='is-active' className="nav-link" to="/" exact={true}>Home</NavLink>
        <NavLink activeClassName='is-active' className="nav-link" to="/notes">Notes</NavLink>
        <NavLink activeClassName='is-active' className="nav-link" to="/about">Organizing your studies</NavLink>
    </header>
)

export default Header;
