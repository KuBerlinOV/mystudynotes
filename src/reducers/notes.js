//reducers = notes
const notesReducerDefaultState = []

const notesReducer = (state = notesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_NOTE':
            return [...state, action.note];
        case 'UPDATE_NOTE':
            return state.map(note => {
                if (note.id === action.id) {
                    return {
                        ...note,
                        ...action.updates
                    }
                } else {
                    return note
                }
            });
        case 'REMOVE_NOTE':
            return state.filter(({ id }) => id !== action.id);
        case 'UPDATE_STATUS':
            return state.map(note => {
                if (note.id === action.id && note.status === 'in progress') {
                    return {
                        ...note,
                        status: 'mastered'
                    }
                } else if (note.id === action.id && note.status === 'mastered') {
                    return {
                        ...note,
                        status: 'in progress'
                    }
                } else {
                    return {
                        ...note
                    }
                }
            })
        default:
            return state;
    }

}

export default notesReducer;
