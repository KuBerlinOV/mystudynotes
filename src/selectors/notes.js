


const selectNotes = (notes, { text, sortBy, startDate, endDate }) => {
    return notes.filter(note => {
        const descriptionToMatch = note.description.toLowerCase()
        const topicToMatch = note.topic.toLowerCase();
        const textToMatch = text.toLowerCase();

        const textMatch = descriptionToMatch.includes(textToMatch) || topicToMatch.includes(textToMatch) ? true : false;
        const startDateMatch = typeof startDate !== 'number' || note.createdAt >= startDate;
        const endDateMatch = typeof endDate !== 'number' || note.createdAt <= endDate;

        return textMatch && startDateMatch && endDateMatch
    }).sort((a, b) => {
        if (sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1
        } else if (sortBy === 'status') {
            return a.status === 'mastered' ? 1 : -1
        }
    })
}

export default selectNotes;