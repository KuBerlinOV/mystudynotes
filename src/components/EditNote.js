import React from 'react';
import { connect } from 'react-redux';
import NoteForm from './NoteForm';
import { updateNote } from '../actions/notes'


const EditNote = (props) => {
    return (
        <div>
            <NoteForm
                //note before update
                noteBeforeUpdate={props.note}
                //this is handle submit props
                handleSubmit={(note) => {
                    //dispatch action to edit the expense
                    props.dispatch(updateNote(props.note.id, note))
                    props.history.push('/notes')
                    //redirect to dashBoard
                }}
                showModal={true} />
        </div>
    )
}

const mapStateToProps = (state, props) => {
    return {
        note: state.notes.find((note) => note.id == props.match.params.id)
    }

}

export default connect(mapStateToProps)(EditNote);