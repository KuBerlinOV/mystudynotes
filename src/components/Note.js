import React, { useState } from 'react';
import { connect } from 'react-redux';
import NoteForm from './NoteForm'
import moment from 'moment'
import { removeNote, updateStatus } from '../actions/notes'
import { Link } from 'react-router-dom';

// note: {
//     id: uuid(),
//     topic,
//     description,
//     note,
//     reference,
//     tag,
//     status,
//     createdAt
// }

export const Note = ({ topic, description, note, reference, tag, createdAt, id, dispatch, status }) => {


    return (
        <div>
            <h3> Topic: {topic} </h3>
            <p>Description: {description} </p>
            <p>Note: {note} </p>
            <p>Reference: {reference} </p>
            <p>Tag: {tag} </p>
            <p>Status: {status}</p>
            <button onClick={() => {
                dispatch(updateStatus(id))
            }}>Change status</button>
            <p>Date: {moment(createdAt).format('MMMM Do YYYY, h:mm')} </p>
            <button onClick={() => {
                dispatch(removeNote(id))
            }
            } >Delete</button>
            <button> <Link to={`/edit/${id}`}> Edit</Link></button>
        </div>
    )
}




export default connect()(Note);