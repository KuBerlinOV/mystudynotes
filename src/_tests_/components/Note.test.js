import React from 'react';
import { shallow } from 'enzyme';
import { Note } from '../../components/Note';
import notes from '../fixtures/notes'

describe('Note', () => {
    it('should render Note', () => {
        const wrapper = shallow(<Note {...notes[0]}
        // topic={notes[0].topic}
        // description={notes[0].description}
        // note={notes[0].note}
        // reference={notes[0].reference}
        // tag={notes[0].tag}
        // createdAt={notes[0].createdAt}
        // id={notes[0].id}
        // status={notes[0].status}
        />)
        expect(wrapper).toMatchSnapshot();
    })
})