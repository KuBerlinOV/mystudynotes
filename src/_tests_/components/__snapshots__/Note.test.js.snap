// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Note should render Note 1`] = `
<div>
  <h3>
     Topic: 
    redux
     
  </h3>
  <p>
    Description: 
    redux note
     
  </p>
  <p>
    Note: 
    here is a note
     
  </p>
  <p>
    Reference: 
    some reference
     
  </p>
  <p>
    Tag: 
    some tag
     
  </p>
  <p>
    Status: 
    mastered
  </p>
  <button
    onClick={[Function]}
  >
    Change status
  </button>
  <p>
    Date: 
    January 1st 1970, 1:00
     
  </p>
  <button
    onClick={[Function]}
  >
    Delete
  </button>
  <button>
     
    <Link
      to="/edit/1"
    >
       Edit
    </Link>
  </button>
</div>
`;
